File: javapatterns/src/main/java/org/iut/nantes/factories/banque/Carte.java
package org.iut.nantes.factories.banque;

abstract class Carte
{
  protected String proprietaire;

  public Carte( String proprietaire )
  {
    this.proprietaire = proprietaire;
  }

  public String getProprietaire() { return proprietaire; }
  
  public abstract String getType();
  public abstract String getPrestation();
}

class Visa extends Carte
{
  public Visa( String proprietaire )
  {
    super( proprietaire );
  }

  public String getType() { return "Visa"; }
  public String getPrestation() { return "assurance annulation avion"; }
}

class MasterCard extends Carte
{
  public MasterCard( String proprietaire )
  {
    super( proprietaire );
  }

  public String getType() { return "MasterCard"; }
  public String getPrestation() { return "assistance juridique à l'étranger"; }
}

class AmericanExpress extends Carte
{
  public AmericanExpress( String proprietaire )
  {
    super( proprietaire );
  }

  public String getType() { return "American Express"; }
  public String getPrestation() { return "réduction de location de voiture"; }
}

File: kotlinpatterns/src/main/kotlin/org/iut/nantes/factories/banque/Carte.kt
package org.iut.nantes.factories.banque

abstract class Carte(var proprietaire: String) {
    abstract val type: String
    abstract val prestation: String
}

class Visa(proprietaire: String) : Carte(proprietaire) {
    override val type: String
        get() = "Visa"
    override val prestation: String
        get() = "assurance annulation avion"
}

class MasterCard(proprietaire: String) : Carte(proprietaire) {
    override val type: String
        get() = "MasterCard"
    override val prestation: String
        get() = "assistance juridique à l'étranger"
}

class AmericanExpress(proprietaire: String) : Carte(proprietaire) {
    override val type: String
        get() = "American Express"
    override val prestation: String
        get() = "réduction de location de voiture"
}